lib/api/api.js

```diff
@@ -184,14 +186,113 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let formDataParser;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            if (request.headers['content-type'].includes('multipart/form-data')) {
+                formDataParser = busboy({ headers: request.headers });
+            } else {
+                const contentTypeError = errors.PreconditionFailed
+                    .customizeDescription('Bucket POST must be of the enclosure-type multipart/form-data');
+                return process.nextTick(callback, contentTypeError);
+            }
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
+                    writeContinue(request, response);
+
+                    request.formData = {};
+                    let totalFieldSize = 0;
+                    const MAX_FIELD_SIZE = 20 * 1024; // 20KB
+                    formDataParser.on('field', (fieldname, val) => {
+                        totalFieldSize += Buffer.byteLength(val, 'utf8');
+                        if (totalFieldSize > MAX_FIELD_SIZE) {
+                            const err = errors.MaxPostPreDataLengthExceeded;
+                            formDataParser.emit('error', err);
+                        }
+                        // Convert fieldname to lowercase for case-insensitive comparison
+                        const lowerFieldname = fieldname.toLowerCase();
+                        request.formData[lowerFieldname] = val;
```

Comment:
> [!CAUTION]
> using `${filename}` in `key` doesn't work
> > To use the file name provided by the user, use the ${filename} variable. For example, if a user named Mary uploads the file example.jpg and you specify /user/mary/${filename}, the key name is /user/mary/example.jpg.

Reply:
I don't understand this comment; what is `${filename}` within the context (`val`?) and what do you mean when you say it doesn't work?

Reply:
This is from the AWS spec in the description of the form field `key`: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html

If the `key` contains the placeholder `${filename}` and the form field `file` has a filename (ie: it has been send as a file and not just text `Content-Disposition: form-data; name="file"; filename="a.txt"
`), this filename needs to replace the placeholder `${filename}` in the `key`.

Right now it's not working as the key keeps the placeholder `${filename}`

Example, this command:
```sh
sed -z 's/\n/\r\n/g' <<"EOF" | nc -N micktestpost.s3.eu-north-1.amazonaws.com 80
POST / HTTP/1.1
Host: micktestpost.s3.eu-north-1.amazonaws.com
Content-Type: multipart/form-data; boundary=abc
Content-Length: 179

--abc
Content-Disposition: form-data; name="key"

THE_KEY_${filename}
--abc
Content-Disposition: form-data; name="file"; filename="this_goes_in_key"

the content
--abc--
EOF
```

Will save the file with a key `THE_KEY_this_goes_in_key` and return

```
HTTP/1.1 204 No Content
Location: http://micktestpost.s3.eu-north-1.amazonaws.com/THE_KEY_this_goes_in_key
...
```

Reply:
Ok I see, thank you. It shouldn't be a problem to do the replacement after policy evaluation, as we would have access to the key from the formData and the file name from the file event data; supposing we maintain a somewhat similar structure after fixing the callback-already-called issue

Reply:
added
https://github.com/scality/cloudserver/pull/5601/files#diff-69284c3576c9c4651c9ab95928cbb676d75009ef4ee5becb238b45e22bb20efeR78

tested
https://github.com/scality/cloudserver/pull/5601/files#diff-406efa736d8942ad5f5ba3c3404803f3748ef0d0cd44a20b48f5ce074c3105cdR686

Reply:
But it's only for the field `file` not any other filename, you should ensure the `fieldname` is actually the `file` to use its name

Reply:
I don't think I understand: the `filename` I'm using here comes directly from the `file` event, and I'm setting that in the `key`, so it is the filename
Am I missing something?

Reply:
see https://github.com/scality/cloudserver/pull/5601#discussion_r1690995359 I'm closing this